Big (hopefully final) ADAR1000 update

wsapia-adi opened the pull request:
- Add missing ADAR1000 attributes
- Add more properties & methods for control of an array as a single entity (attenuators, gains, phases, steering, etc.)
- Nest adar1000_channel class under adar1000 to allow for simpler access to the elements in the array (can be accessed via element number in the array class, or from the parent ADAR1000 object)
- Add support for LTC2314-14
- Add LTC2314-14 and one_bit_adc_dac to the documentation

tfcollins commented:
Sorry for the delay. Got the go-ahead from Mircea.

1. Squash the formatting/cleanup commits
2. Doc is generating warnings
```
eading sources... [100%] support                                                                                     
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:21: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:25: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000:32: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:19: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:21: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:26: WARNING: Unexpected indentation.
/tmp/pyadi-iio-adar1000/adi/adar1000.py:docstring of adi.adar1000.adar1000_array:30: WARNING: Block quote ends without a blank line; unexpected unindent.
/tmp/pyadi-iio-adar1000/adi/one_bit_adc_dac.py:docstring of adi.one_bit_adc_dac.one_bit_adc_dac:4: WARNING: Unexpected indentation.
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
```
3. For the singleton adar1000 class, can you add a note to each input parameter in regards to which are needed only for the array implementation. It can be a bit overwhelming if you just have a single eval board

tfcollins commented on adi/adar1000.py:
```diff
@@ -41,31 +43,374 @@ class adar1000(attribute, context_manager):
+    _BIAS_CODE_TO_VOLTAGE_SCALE = -0.018824
+
+    class adar1000_channel:
+        """Class for each channel of the ADAR1000. This class is not meant
```
Due to the instantiation comment can you add a check in __init__ like: 
```
if type(self) == adar1000_channel:
    raise something
```

wsapia-adi replied:
I can add this, but what do you think about just hiding it with a leading "_"?

tfcollins replied:
Added a _ prefix is fine.

wsapia-adi replied:
I should undo that then. It needs to show up in the doc so that people can see what properties/methods are available.

wsapia-adi replied:
I'm not sure I can add your check that you originally suggested, as I believe that will always fail? During __init__, won't self always equal adar1000_channel?

tfcollins commented on adi/adar1000.py:
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
+
+        # Latch in the new phases
+        self.latch_tx_settings()
+        return
```
Remove

wsapia-adi replied:
Why remove all of these here and below? If these aren't there, none of the new values will take effect...

tfcollins replied:
What is the purpose of the return statement?

wsapia-adi replied:
I didn't realize you were referring to the return, I thought you were saying to remove the latch command. I get that it doesn't return anything, but it works the same either way. Just a habit. I can remove it.

tfcollins commented on adi/adar1000.py:
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
+
+        # Latch in the new phases
+        self.latch_rx_settings()
+        return
```
Remove

tfcollins commented on adi/adar1000.py:
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
+        """ Latch in new Gain/Phase settings for the Tx """
+        for device in self.devices.values():
+            device.latch_tx_settings()
+        return
```
Remove

tfcollins commented on adi/adar1000.py:
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
+        """ Latch in new Gain/Phase settings for the Rx """
+        for device in self.devices.values():
+            device.latch_rx_settings()
+        return
```
Remove

tfcollins commented on adi/adar1000.py:
```diff
@@ -781,102 +1257,431 @@ class adar1000_array(context_manager):
+        """ Initialize the ADAR1000s in the array """
+        for device in self.devices.values():
+            device.initialize()
+        return
```
Remove

wsapia-adi commented:
1. Tried to squash all the formatting and miscellaneous commits, let me know if this is clean enough
2. Not sure what's causing this?
3. Added more info to the docstrings, clearer?
4. Added example files for single and array instances

tfcollins commented on adi/adar1000.py:
```diff
@@ -85,693 +440,820 @@ def __init__(self, uri="", chip_id="csb1_chip1", context_handle=None):
+        # Latch in the new settings
+        self.latch_rx_settings()
+        self.latch_tx_settings()
+        return
```
Remove. I think that's the last one and we should be good to go.
